core/include/cubos/core/ecs/event_reader.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef CUBOS_CORE_ECS_EVENT_READER_HPP
+#define CUBOS_CORE_ECS_EVENT_READER_HPP
+
+#include <cubos/core/ecs/event_pipe.hpp>
+#include <optional>
+
+namespace cubos::core::ecs
+{
+    /// Used to read events, filtering is also possible via M parameter.
+    /// In case M template parameter is not provided, it will read all events (no filtering).
+    /// @tparam T Event.
+    /// @tparam M Event mask.
+    template <typename T, unsigned int M = 0> class EventReader
+    {
+    public:
+        EventReader(EventPipe<T>& pipe);
+
+        /// Returns a reference to current event, and advances.
+        /// It returns nullopt if there are no more events to read!
+        std::optional<std::reference_wrapper<T>> read();
+
+        /// EventReader custom iterator.
+        class Iterator
+        {
+        public:
+            Iterator(EventReader<T, M>& reader, std::optional<std::reference_wrapper<T>> ev, bool end);
+
+            T& operator*();
+            Iterator& operator++();
+            bool operator==(const Iterator& other);
+            bool operator!=(const Iterator& other);
+
+        private:
+            EventReader<T, M>& reader;
+            std::optional<std::reference_wrapper<T>> currentEvent;
+            bool end;
+        };
+
+        Iterator begin();
+        Iterator end();
+
+    private:
+        EventPipe<T>& pipe;
+        std::size_t index{};
+
+        /// Checks if given mask is valid to reader's one.
+        /// @return True if mask is valid.
+        bool matchesMask(decltype(M) mask) const;
+    };
+
+    // EventReader implementation.
+
+    template <typename T, unsigned int M> EventReader<T, M>::EventReader(EventPipe<T>& pipe) : pipe(pipe)
+    {
+        this->pipe.addReader();
+    }
+
+    template <typename T, unsigned int M> std::optional<std::reference_wrapper<T>> EventReader<T, M>::read()
+    {
+        while (this->index < this->pipe.size())
+        {
+            auto [event, mask] = pipe.get(this->index++);
+            if (matchesMask(mask))
+            {
+                return event;
+            }
+        }
+
+        return std::nullopt;
+    }
+
+    template <typename T, unsigned int M> bool EventReader<T, M>::matchesMask(decltype(M) eventMask) const
+    {
+        if (M == 0)
+        {
+            return true;
+        }
+
+        if (!(eventMask & M) || (eventMask & M) != eventMask)
+        {
+            return false;
+        }
+
+        return true;
```

Comment:
Well, with this condition, events with mask `0` would never be read, since no bits are set. So,we have two choices:
- either we keep it as is, events with mask 0 are never read, and we change the default push mask to `~0`.
- or we make it such that events with `mask 0` are read only when `M = ~0` (e.g., when no filter is specified)

Reply:
I think im going with the first approach, and maybe add a `static_assert` (compile-time) for forcing `M` to be != 0... Since it makes no sense to use masking events with `mask=0`